drivers/net/ppp.c

```diff
@@ -0,0 +1,633 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * PPP driver using uart_pipe. This is meant for network connectivity between
+ * two network end points.
+ */
+
+#define LOG_LEVEL CONFIG_NET_PPP_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_ppp, LOG_LEVEL);
+
+#include <stdio.h>
+
+#include <kernel.h>
+
+#include <stdbool.h>
+#include <errno.h>
+#include <stddef.h>
+#include <net/ppp.h>
+#include <net/buf.h>
+#include <net/net_pkt.h>
+#include <net/net_if.h>
+#include <net/net_core.h>
+#include <console/uart_pipe.h>
+#include <crc.h>
+
+#include "../../subsys/net/ip/net_stats.h"
+#include "../../subsys/net/ip/net_private.h"
+
+#define UART_BUF_LEN CONFIG_NET_PPP_UART_PIPE_BUF_LEN
+
+enum ppp_driver_state {
+	STATE_HDLC_FRAME_START,
+	STATE_HDLC_FRAME_ADDRESS,
+	STATE_HDLC_FRAME_DATA,
+};
+
+struct ppp_driver_context {
+	struct net_if *iface;
+
+	/* This net_pkt contains pkt that is being read */
+	struct net_pkt *pkt;
+
+	/* How much free space we have in the net_pkt */
+	size_t available;
+
+	/* ppp data is read into this buf */
+	u8_t buf[UART_BUF_LEN];
+
+	/* ppp buf use when sending data */
+	u8_t send_buf[UART_BUF_LEN];
+
+	u8_t mac_addr[6];
+	struct net_linkaddr ll_addr;
+
+#if defined(CONFIG_NET_STATISTICS_PPP)
+	struct net_stats_ppp stats;
+#endif
+
+	enum ppp_driver_state state;
+
+	u8_t init_done : 1;
+	u8_t next_escaped : 1;
+};
+
+static struct ppp_driver_context ppp_driver_context_data;
+
+static int ppp_save_byte(struct ppp_driver_context *ppp, u8_t byte)
+{
+	int ret;
+
+	if (!ppp->pkt) {
+		ppp->pkt = net_pkt_rx_alloc_with_buffer(
+			ppp->iface,
+			CONFIG_NET_BUF_DATA_SIZE,
+			AF_UNSPEC, 0, K_NO_WAIT);
+		if (!ppp->pkt) {
+			LOG_ERR("[%p] cannot allocate pkt", ppp);
+			return -ENOMEM;
+		}
+
+		net_pkt_cursor_init(ppp->pkt);
+
+		ppp->available = net_pkt_available_buffer(ppp->pkt);
+	}
+
+	/* Extra debugging can be enabled separately if really
+	 * needed. Normally it would just print too much data.
+	 */
+	if (0) {
+		LOG_DBG("Saving byte %02x", byte);
+	}
+
+	/* This is not very intuitive but we must allocate new buffer
```

Comment:
Hmm, I prefer leaving this comment here for time being. I spent hours debugging this as it was not clear why the api requires that. At least for me, it would be difficult to understand why we have the ppp->available==1 check without the comment. The documentation of net-pkt should be improved in this respect.